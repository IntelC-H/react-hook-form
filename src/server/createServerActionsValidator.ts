import {
  DeepMap,
  DeepPartial,
  ErrorOption,
  FieldErrors,
  FieldPath,
  FieldValues,
  RegisterOptions,
  Resolver,
} from '../types';
import set from '../utils/set';

class ServerActionsValidator<TFieldValues extends FieldValues> {
  private values: TFieldValues;
  private options: { resolver: Resolver<TFieldValues> };
  private rules: DeepMap<
    DeepPartial<TFieldValues>,
    RegisterOptions<TFieldValues>
  > = {} as DeepMap<DeepPartial<TFieldValues>, RegisterOptions<TFieldValues>>;
  private errors: FieldErrors<TFieldValues> = {};

  constructor(
    values: TFieldValues | FormData,
    options: { resolver: Resolver<TFieldValues> },
  ) {
    this.values =
      values instanceof FormData ? this.formDataToValues(values) : values;
    this.options = options;
  }

  private formDataToValues(formData: FormData) {
    return Object.fromEntries(formData.entries()) as TFieldValues;
  }

  public register(
    name: FieldPath<TFieldValues>,
    options: RegisterOptions<TFieldValues>,
  ) {
    set(this.rules, name, options);
    return this;
  }

  public async validate() {
    if (this.options.resolver) {
      const result = await this.options.resolver(this.values, undefined, {
        fields: {},
        shouldUseNativeValidation: false,
      });
      this.errors = {
        ...this.errors,
        ...result.errors,
      };
      return this;
    }
    // TODO: impl native validation
    return this;
  }

  public setError(
    name: FieldPath<TFieldValues> | `root.${string}` | 'root',
    error: ErrorOption,
  ) {
    set(this.errors, name, error);
    return this;
  }

  public getResult(): {
    values: TFieldValues;
    errors: FieldErrors<TFieldValues>;
  } {
    return {
      values: this.values,
      errors: this.errors,
    };
  }

  public isValid() {
    return Object.keys(this.errors).length === 0;
  }
}

export const createServerActionsValidator = <TFieldValues extends FieldValues>(
  values: TFieldValues | FormData,
  options: { resolver: Resolver<TFieldValues> },
) => new ServerActionsValidator(values, options);
