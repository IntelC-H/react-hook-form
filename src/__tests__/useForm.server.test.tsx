import * as React from 'react';
import { renderToString } from 'react-dom/server';

import { useForm } from '../useForm';

describe('useForm with SSR', () => {
  it('should not output error', () => {
    const Component = () => {
      const { register } = useForm<{
        test: string;
      }>();
      return (
        <div>
          <input {...register('test')} />
        </div>
      );
    };

    const spy = jest.spyOn(console, 'error');

    expect(renderToString(<Component />)).toMatchSnapshot();

    expect(spy).not.toHaveBeenCalled();
  });

  it('should not pass down constrained API for server side rendering', () => {
    const App = () => {
      const { register } = useForm<{
        test: string;
      }>();

      return (
        <div>
          <input
            {...register('test', {
              required: true,
              min: 2,
              max: 2,
              maxLength: 2,
              minLength: 2,
            })}
          />
        </div>
      );
    };

    expect(renderToString(<App />)).toMatchSnapshot();
  });

  it('should pass down constrained API for server side rendering', () => {
    const App = () => {
      const { register } = useForm<{
        test: string;
      }>({
        shouldUseNativeValidation: true,
      });

      return (
        <div>
          <input
            {...register('test', {
              required: true,
              min: 2,
              max: 2,
              maxLength: 2,
              minLength: 2,
            })}
          />
        </div>
      );
    };

    expect(renderToString(<App />)).toMatchSnapshot();
  });
});
