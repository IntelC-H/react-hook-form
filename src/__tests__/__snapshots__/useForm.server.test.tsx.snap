// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useForm with SSR should not output error 1`] = `"<div data-reactroot=\\"\\"><input name=\\"test\\"/></div>"`;

exports[`useForm with SSR should not pass down constrained API for server side rendering 1`] = `"<div data-reactroot=\\"\\"><input name=\\"test\\"/></div>"`;

exports[`useForm with SSR should pass down constrained API for server side rendering 1`] = `"<div data-reactroot=\\"\\"><input required=\\"\\" min=\\"2\\" max=\\"2\\" minLength=\\"2\\" maxLength=\\"2\\" name=\\"test\\"/></div>"`;
